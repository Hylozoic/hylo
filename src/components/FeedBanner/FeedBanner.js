import React from 'react'
import cx from 'classnames'
import { Link } from 'react-router-dom'
import { bgImageStyle } from 'util/index'
import { DEFAULT_BANNER, DEFAULT_AVATAR } from 'store/models/Group'
import './FeedBanner.scss'
import { whiteMerkaba, allGroupsBanner, publicGlobe } from 'util/assets'
import Icon from 'components/Icon'
import RoundImage from 'components/RoundImage'

export default function FeedBanner ({
  context,
  group,
  currentUser,
  newPost,
  type,
  urlLocation,
  currentUserHasMemberships
}) {
  let bannerUrl, avatarUrl, name, location, subtitle

  if (context === 'all') {
    name = 'All My Groups'
    avatarUrl = whiteMerkaba
    bannerUrl = allGroupsBanner
    subtitle = currentUser && `${currentUser.memberships.count()} Groups`
  } else if (context === 'public') {
    name = 'Public Groups & Posts'
    avatarUrl = publicGlobe
    bannerUrl = allGroupsBanner
    // TODO list count of public posts and public groups in subtitle
    subtitle = `All Posts Marked Public`
  } else if (!group) {
    return null
  } else {
    ({ bannerUrl, avatarUrl, name, location } = group)
  }

  return <div styleName={cx('banner', { 'all-groups': context === 'all' })}>
    <div style={bgImageStyle(bannerUrl || DEFAULT_BANNER)} styleName='image'>
      <div styleName='fade'><div styleName='fade2' /></div>
      <div styleName='header'>
        <div styleName={cx('logo', { 'all-logo': context === 'all' })} style={bgImageStyle(avatarUrl || DEFAULT_AVATAR)} />
        <div styleName='header-text'>
          <div styleName='header-contents'>
            <span styleName='header-name'>{name}</span>
            {location && <div styleName='header-subtitle'>
              <Icon name='Location' styleName='header-icon' />
              {location}
            </div>}
            {subtitle && <div styleName='header-subtitle'>
              {subtitle}
            </div>}
          </div>
        </div>
      </div>
    </div>
    {currentUserHasMemberships && <PostPrompt
      type={type}
      location={urlLocation}
      firstName={currentUser.firstName()}
      avatarUrl={currentUser.avatarUrl}
      newPost={newPost} />}
  </div>
}

export function postPromptString (type = '', { firstName }) {
  const postPrompts = {
    offer: `Hi ${firstName}, what would you like to share?`,
    request: `Hi ${firstName}, what are you looking for?`,
    project: `Hi ${firstName}, what would you like to create?`,
    event: `Hi ${firstName}, want to create an event?`,
    default: `Hi ${firstName}, what's on your mind?`
  }

  return postPrompts[type] || postPrompts['default']
}

export class PostPrompt extends React.Component {
  static defaultProps = {
    type: '',
    location: '',
    firstName: '',
    promptStringFunc: postPromptString
  }

  constructor (props) {
    super(props)
    this.state = { hover: false }
  }

  onMouseEnterHandler = () => this.setState({ hover: true })

  onMouseLeaveHandler = () => this.setState({ hover: false })

  render () {
    const { type, location, avatarUrl, firstName, promptStringFunc, className } = this.props
    const { hover } = this.state

    return <div onMouseEnter={this.onMouseEnterHandler} onMouseLeave={this.onMouseLeaveHandler}>
      <Link to={location.pathname + '/create/post?newPostType=' + type}>
        <div styleName='postPrompt' className={className}>
          <RoundImage url={avatarUrl} small styleName='prompt-image' />
          {promptStringFunc(type, { firstName })}
        </div>
      </Link>
      <div styleName={cx('shadow', { hover })} />
    </div>
  }
}
