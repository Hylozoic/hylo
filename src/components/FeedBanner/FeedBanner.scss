.banner {
  padding-bottom: $space-8x;
  position: relative;
  z-index: 20;
}

.imageShape {
  width: 100%;
  height: 160px;
  border-radius: 0 0 10px 10px;
}

.image {
  composes: imageShape;
  background-position: center;
  background-size: cover;
  overflow: hidden;
}

.fade {
  // Note: replace these with pseudo elements
  composes: imageShape;
  position: relative;
}

.fade2 {
  composes: imageShape;
  background: linear-gradient(180deg, #2A4059 0%, rgba(42, 64, 89, 0.2) 0%, rgba(42, 64, 89, 0.5) 100%);
}

.header {
  position: relative;
  top: -110px;
  padding-left: 20px;
  display: flex;
}

.header-contents {
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
}

.logo {
  min-width: 4rem;
  min-height: 4rem;
  height: 4rem;
  width: 4rem;
  background-size: cover;
  border-radius: 100%;
  margin-right: 10px;
  border: 3px solid rgba(255, 255, 255, 1);
}

.all-logo {
  width: 55px;
  height: 62px;
}

.header-text {
  display: inline-block;
  vertical-align: top;
  line-height: 1.1;
}

.header-name {
  // Extract typography?
  @include font-medium;
  color: $color-white;
  font-size: 24px;
  letter-spacing: -0.47px;
  text-shadow: 0 2px 4px rgba(0,0,0,0.50);
}

.header-subtitle {
  @include font-regular;
  color: $color-white;
  font-size: 12px;
  opacity: 0.7;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 80%;
}

.header-icon {
  position: relative;
  top: 1px;
  margin-right: 5px;
  font-size: 16px;
  color: $color-white;
}

// PostPrompt

// Note: find other places where 56px and 28px are used.
.postPromptShape {
  border-radius: 4px;
  height: 56px;
  position: absolute;
  bottom: $space-3x;
  left: $space-4x;
  right: $space-4x;
}

.postPrompt {
  composes: postPromptShape;
  background-color: $color-white;
  line-height: 56px;
  margin: 0 auto;
  cursor: pointer;
  @include font-regular;
  font-size: 16px;
  color: $color-rhino-30;
  text-decoration: none;
}

.prompt-image {
  margin-right: 10px;
  margin-left: 20px;
  position: relative;
  top: 14px;
}

.shadow {
  composes: postPromptShape;
  background-color: transparent;
  box-shadow: 0px 4px 42px 0px darken($color-seashell, 10%);
  z-index: -10;
  &.hover {
    box-shadow: 0px 4px 42px 0px darken($color-seashell, 30%);
  }
}

.all-groups {
  background-color: $color-athens-gray;
  .header-name {
    text-shadow: none;
  }
  .fade {
    background-image: none;
  }
  .fade2 {
    display: none;
  }
}

@media screen and (max-width: 600px) {
  .imageShape {
    border-radius: 0;
  }

  .banner {
    padding-top: 0;
  }
}

@media screen and (max-width: 425px) {
  .header {
    padding-left: 15px;
  }

  .header-name {
    font-size: 18px;
  }

  .postPrompt {
    font-size: 14px;
    line-height: 50px;
  }

  .postPromptShape {
    height: 50px;
  }

  .prompt-image {
    margin-left: 10px;
    top: 11px;
  }
}

@media screen and (max-width: 325px) {
  .header-name {
    font-size: 16px;
  }

  .postPrompt {
    font-size: 12px;
  }
}
