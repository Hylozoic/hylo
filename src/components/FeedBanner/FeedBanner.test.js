import React from 'react'
import FeedBanner, { postPromptString } from './FeedBanner'
import { mount } from 'enzyme'

const currentUser = {
  avatarUrl: 'me.png',
  firstName: () => 'Bob',
  memberships: {
    count: () => 18
  }
}

const group = {
  bannerUrl: 'banner.png',
  avatarUrl: 'avatar.png',
  name: 'Spacebase',
  location: 'space, duh'
}

it('renders with a group', () => {
  const node = mount(<FeedBanner
    group={group}
    isTesting
  />)
  expect(node).toMatchSnapshot()
})

it('renders for all groups', () => {
  const node = mount(<FeedBanner
    all
    currentUser={currentUser}
    currentUserHasMemberships
  />)
  expect(node).toMatchSnapshot()
})

it('matches the snapshot for an orphan user', () => {
  const node = mount(<FeedBanner
    all
    currentUser={currentUser}
    currentUserHasMemberships={false}
  />)
  expect(node).toMatchSnapshot()
})

describe('postPromptString', () => {
  it('renders a post prompt string', () => {
    const firstName = 'anybody'

    expect(postPromptString('project', { firstName })).toMatchSnapshot()
    expect(postPromptString('', { firstName })).toMatchSnapshot()
  })
})
