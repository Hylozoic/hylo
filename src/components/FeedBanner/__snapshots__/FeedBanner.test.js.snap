// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot for an orphan user 1`] = `
<FeedBanner
  all={true}
  currentUser={
    Object {
      "avatarUrl": "me.png",
      "firstName": [Function],
      "memberships": Object {
        "count": [Function],
      },
    }
  }
  currentUserHasMemberships={false}
/>
`;

exports[`postPromptString renders a post prompt string 1`] = `"Hi anybody, what would you like to create?"`;

exports[`postPromptString renders a post prompt string 2`] = `"Hi anybody, what's on your mind?"`;

exports[`renders for all groups 1`] = `
<FeedBanner
  all={true}
  currentUser={
    Object {
      "avatarUrl": "me.png",
      "firstName": [Function],
      "memberships": Object {
        "count": [Function],
      },
    }
  }
  currentUserHasMemberships={true}
/>
`;

exports[`renders with a group 1`] = `
<FeedBanner
  group={
    Object {
      "avatarUrl": "avatar.png",
      "bannerUrl": "banner.png",
      "location": "space, duh",
      "name": "Spacebase",
    }
  }
>
  <div
    data-stylename="banner"
  >
    <div
      data-stylename="image"
      style={
        Object {
          "backgroundImage": "url(banner.png)",
        }
      }
    >
      <div
        data-stylename="fade"
      >
        <div
          data-stylename="fade2"
        />
      </div>
      <div
        data-stylename="header"
      >
        <div
          data-stylename="logo"
          style={
            Object {
              "backgroundImage": "url(avatar.png)",
            }
          }
        />
        <div
          data-stylename="header-text"
        >
          <div
            data-stylename="header-contents"
          >
            <span
              data-stylename="header-name"
            >
              Spacebase
            </span>
            <div
              data-stylename="header-subtitle"
            >
              <Icon
                data-stylename="header-icon"
                name="Location"
              >
                <span
                  className="icon-Location"
                  data-stylename="icon"
                />
              </Icon>
              space, duh
            </div>
          </div>
        </div>
      </div>
    </div>
    <ReactTooltip
      backgroundColor="rgba(35, 65, 91, 1.0)"
      clickable={false}
      delayShow={0}
      effect="solid"
      getContent={[Function]}
      id="feed-banner-tip"
      insecure={true}
      place="bottom"
      resizeHide={true}
      wrapper="div"
    >
      <div
        className="__react_component_tooltip ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad place-bottom type-dark"
        data-id="tooltip"
        id="feed-banner-tip"
      >
        <style
          aria-hidden="true"
          dangerouslySetInnerHTML={
            Object {
              "__html": "
  	.ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad {
	    color: #fff;
	    background: #222;
	    border: 1px solid transparent;
  	}

  	.ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-top {
        margin-top: -10px;
    }
    .ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-top::before {
        border-top: 8px solid transparent;
    }
    .ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-top::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        bottom: -6px;
        left: 50%;
        margin-left: -8px;
        border-top-color: #222;
        border-top-style: solid;
        border-top-width: 6px;
    }

    .ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-bottom {
        margin-top: 10px;
    }
    .ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-bottom::before {
        border-bottom: 8px solid transparent;
    }
    .ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-bottom::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        top: -6px;
        left: 50%;
        margin-left: -8px;
        border-bottom-color: #222;
        border-bottom-style: solid;
        border-bottom-width: 6px;
    }

    .ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-left {
        margin-left: -10px;
    }
    .ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-left::before {
        border-left: 8px solid transparent;
    }
    .ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-left::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        right: -6px;
        top: 50%;
        margin-top: -4px;
        border-left-color: #222;
        border-left-style: solid;
        border-left-width: 6px;
    }

    .ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-right {
        margin-left: 10px;
    }
    .ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-right::before {
        border-right: 8px solid transparent;
    }
    .ta20a7e52-6ffb-4e0b-8df0-c19ba10f8bad.place-right::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        left: -6px;
        top: 50%;
        margin-top: -4px;
        border-right-color: #222;
        border-right-style: solid;
        border-right-width: 6px;
    }
  ",
            }
          }
        />
      </div>
    </ReactTooltip>
  </div>
</FeedBanner>
`;
